File: python-dist/src/qntm/cbor.py
"""Canonical CBOR encoding/decoding for QSP.

RFC 8949 section 4.2.1: deterministic encoding with keys sorted
by encoded byte length first, then lexicographic.
"""

import cbor2


def _to_canonical(obj):
    """Recursively convert dicts to sorted maps for canonical CBOR."""
    if obj is None:
        return obj
    if isinstance(obj, bytes):
        return obj
    if isinstance(obj, memoryview):
        return bytes(obj)
    if isinstance(obj, list):
        return [_to_canonical(item) for item in obj]
    if isinstance(obj, dict):
        # RFC 8949 §4.2.1: sort by encoded key length first, then lexicographic
        keys = sorted(obj.keys(), key=lambda k: (len(k), k))
        return {k: _to_canonical(obj[k]) for k in keys}
    return obj


def marshal_canonical(value) -> bytes:
    """Encode value as canonical CBOR."""
    canonical = _to_canonical(value)
    return cbor2.dumps(canonical, canonical=True)


def unmarshal(data: bytes):
    """Decode CBOR bytes."""
    return cbor2.loads(data)

File: client/src/crypto/cbor.ts
import { Encoder, Decoder } from 'cbor-x';

const encoder = new Encoder({
  useRecords: false,
  mapsAsObjects: false,
  tagUint8Array: false,
} as any);

const decoder = new Decoder({
  useRecords: false,
  mapsAsObjects: true,
} as any);

/**
 * Sort object keys for canonical CBOR (RFC 8949 §4.2.1).
 * Keys sorted by encoded byte length first, then lexicographic.
 * Convert objects to Maps so cbor-x uses minimal map headers.
 */
function toCanonicalMap(obj: unknown): unknown {
  if (obj === null || obj === undefined) return obj;
  if (obj instanceof Uint8Array || Buffer.isBuffer(obj)) return obj;
  if (Array.isArray(obj)) return obj.map(toCanonicalMap);
  if (typeof obj === 'object') {
    const keys = Object.keys(obj as Record<string, unknown>);
    // RFC 8949 §4.2.1: sort by encoded key bytes (length-first for CBOR strings)
    keys.sort((a, b) => {
      if (a.length !== b.length) return a.length - b.length;
      return a < b ? -1 : a > b ? 1 : 0;
    });
    const map = new Map<string, unknown>();
    for (const key of keys) {
      map.set(key, toCanonicalMap((obj as Record<string, unknown>)[key]));
    }
    return map;
  }
  return obj;
}

export function marshalCanonical(value: unknown): Uint8Array {
  const canonical = toCanonicalMap(value);
  return encoder.encode(canonical);
}

export function unmarshalCanonical<T = unknown>(data: Uint8Array): T {
  return decoder.decode(data) as T;
}
